lib/Twig/Extension/Core.php

```diff
@@ -821,10 +821,32 @@ function _twig_default_filter($value, $default = '')
 function twig_get_array_keys_filter($array)
 {
     if ($array instanceof Traversable) {
-        return array_keys(iterator_to_array($array));
+        if ($array instanceof IteratorAggregate) {
+            $array = $array->getIterator();
+        } // else $array must be an instance of IteratorAggregate here, we could call getIterator on that one etc. until we get an iterator
```

Comment:
Removed.

What I was thinking of when making the note is that when we know the object being an instance of `Traversable` we can always get the `Iterator`. Now we only try it one time with `getIterator()`.
We could do something like:
```php
if ($array instanceof Traversable) { // so either implements IteratorAggregate or Iterator.
    if ($array instanceof IteratorAggregate) {
        do {
            // returns either a IteratorAggregate or Iterator.
            $array = $array->getIterator(); 
        } while($array instanceof IteratorAggregate);
    }

    // $array is always an instance of Iterator here
}
```